### ISSUE TITLE
[icon library] - add react icon names and download functionality to icon cards

### ISSUE THREAD

**Issue by jnm2377** [carbon-design-system/carbon-website#766]
### Summary

Icon library page currently only shows the regular name for icons, but we don't have anywhere that lists icons according to their `icons-react` name. i.e. `Misuse24` `CheckmarkFilled24`, etc.

### Desired outcome

overflow menu allowing "export of icons"
- raw svg
- download svg
- react component

cc: @joshblack

**Comment by vpicone** [comment]
I'll take this, the hold up on this previously was how we want to display it: tooltip vs button that copies to clipboard etc.

**Comment by vpicone** [comment]
Ideally we'd have an overflow menu that would present:
- React
- SVG Download
- Angular maybe?


We could also have the icons themselves be clickable to copy the React code (our preferred implementation)

**Comment by vpicone** [comment]
@joshblack which size would it make sense to have copied here do you think? Maybe this would be a good time to implement the `<Checkmark size='sm' />` api.

**Comment by joshblack** [comment]
@vpicone my hunch would be 16 or 20, icons in a broader sense should be revisited so that we could do things like what you're suggesting.

If we want to, we could expose the module base name (like `CheckmarkFilled`) and share info about the sizes, but I totally understand if that'd be confusing.

For icon packages, most likely we'll want to only expose React as Angular/Vue do not have active attention dedicated to them (in fact Angular hasn't been published in the previous minors due to build issues)

**Comment by vpicone** [comment]
@joshblack that makes sense, I think the only other thing we'd want is a `download` menu item then @laurenmrice

**Comment by chrisconnors-ibm** [comment]
We do not want to provide a way to easily download SVGs from the icon page. This leads to orphaned icons in product endpoints that never catch iterations or updates.

**Comment by joshblack** [comment]
How did we want to handle SVG downloads for presentations? I think that's one case that did come up recently for wanting to download them.

**Comment by chrisconnors-ibm** [comment]
We could provide code snippets to call them from our packages, and maybe their menu path in the Sketch kit, but not do things that make it super easy for teams to do things we actively should be discouraging (SVG downloads).

> How did we want to handle SVG downloads for presentations? I think that's one case that did come up recently for wanting to download them.

This is more of a pictograms use case, so maybe we do something different there. but the net is for a preso you can get them all from the repo as part of the package or you can dl the illustrator master file.

**Comment by chrisconnors-ibm** [comment]
if you need just one from the repo, the raw button gets you https://raw.githubusercontent.com/carbon-design-system/carbon/master/packages/icons/svg/32/badge.svg

**Comment by chrisconnors-ibm** [comment]
to summarize - no raw or download svg from the icon cards or pages please

**Comment by mjabbink** [comment]
I think this is a must for pictograms. It’s  a weekly request and even from my boss Terry and some other executives. They do not want to poke around in GitHub.

I’s also extend this functionality to UI icons as the same request comes up there as well but less frequent. Not all people use Sketch or even understand GitHub for that matter but want to use and have access to individual icons.

Holding them sacred also pushes people to do worse shit that I’ve seen equally or more distasteful.

Let’s please proceed with this. I understand the downsides but the entire team and a very large number of users will appreciate this.

**Comment by mjabbink** [comment]
So if i wan’t clear. Please execute this issue.

**Comment by chrisconnors-ibm** [comment]
If we do this for _icons_ we will _encourage_ people to do the easy thing (grab the SVG/DL the icon) rather than the _right thing_ (access our icons via the packages, or the design kits) and the FIRST PERSON to notice and complain about teams using an icon we've iterated on in their applications (think: copy icon) will be you, @mjabbink. And you're going to be complaining a lot more then because they will never, ever, track down that SVG code and fix it.

We've already (and always) said "make direct DL possible for pictograms, because orphaning them doesn't matter", so please by all means add direct dl to these cards: https://www.carbondesignsystem.com/guidelines/pictograms/library 

…but that should probably be a separate issue. Up to you all.

I suspect @designertyler has a spec for these types of functionality.

**Comment by chrisconnors-ibm** [comment]
We should add a link somewhere to the .ai master file in the repo: https://github.com/carbon-design-system/carbon/blob/master/packages/icons/master/ui-icon-master.ai

**Comment by mjabbink** [comment]
I’m more inclined to this for Pictograms. I’m not so sure holding back on icons is a real preventative measure against poor behavior. I also do not foresee adjustments to icons. I see new ones being added as we do plenty of that but date we have not modified many icons over the past year. Unless you consider the 16px ones you and Conrad did. I would be fine with first releasing this functionality with pictograms then icons later to test the waters.

**Comment by aagonzales** [comment]
While encouraging people to use icons through the Sketch library works if you're a Sketch user its starting not to scale as we introduce other tools, like Axure and Adobe XD. Right now the icons aren't available in those tools and accessing them from the website is easier than having to have a second tool (Sketch) or getting them from GitHub which many designers are comfortable with outside of issue creation. Sketch licenses are also hard to come by outside of design.

Not using the icon symbol in Sketch is the same as detaching a symbol, it only really hurts the person that does it. I think accommodating for these other usages by adding the download to the website will be more helpful than worrying about people not using the Sketch library.

**Comment by designertyler** [comment]
I tried to summarize the requirements mentioned here so we can generate design specs. Let me know if we're missing any or if some should be a phase 1 or phase 2, 3, 4, etc. 

**Requirement 1: Add react icon names to icon card**
As a product developer want to see the different sized React icon names so I can copy and paste the correct size into my project.


**Requirement 1.2: Add angular icon names to icon card (?)**
As a product developer want to see the different sized Angular icon names so I can copy and paste the correct size into my project.


**Requirement 2: Enable direct download for pictograms**
As a product designer I want to download the SVG asset of a specific pictogram so I can use it in my project.


**Requirement 2.2: Enable bulk download for pictograms**
As a product designer I want to select multiple Pictogram SVG's and download all at once so I don't have to repeat the same action to get all the pictograms for my project.


**Requirement 3: Inform the preferred method to get icon assets**
As a product designer I can learn how to get the icon asset in Sketch without directly downloading the SVG.


**Requirement 4: Enable direct download for icons**
As a product designer I want to download the SVG asset of a specific icon so I can use it in my project.

**Requirement 4.2: Enable bulk download for icons**
As a product designer I want to select multiple icon SVG's and download all at once so I don't have to repeat the same action to get all the pictograms for my project.

**Comment by mjabbink** [comment]
Please proceed with issue defined above. Meaning direct download and copy for both icons and Pictograms

**Comment by joshblack** [comment]
Just a heads up @vpicone I'm updating our icon metadata structure to make it easier to parse for different meta fields, let me know if you want to sync up on this work 👍

**Comment by vpicone** [comment]
@joshblack sounds good, a list of changes would suffice! we do the flattening of the metadata ourselves at run time at the moment.

@chrisconnors-ibm Downloading svg's is a real use case, linking to the master file isn't a solution for things like keynote/powerpoint/non-sketch design tools.

Not sure why pictograms and icons would be treated different. I've always felt the bespoke asset per size icon routine was a mistake. SVG = **scaleable** vector graphic. We're using them like sprites.

**Comment by vpicone** [comment]
@designertyler Requirement 4 and 5 seem the same. Icons and Pictograms are treated very similarly by the website so realistically things would be added in tandem.

**Comment by vpicone** [comment]
@joshblack the biggest blocker/hurdle for me is the sizes in the icon names at this point. I feel like a bit of a broken record for saying this, but these are scaleable vector graphics, they shouldn’t be bespoke sprites at individual sizes.

Having the copied code read `Chevron16` or `Chevron24` and making an assumption on which size they want (**and the assuming that they’d even know what that number means or what the other options are**) is untenable to me.

**Comment by joshblack** [comment]
@vpicone I'm pretty sure we would need to constrain icon size no matter what, e.g. some assets will need to be used in a 16px context or some in a 20px context. This was the basis for the icon size tokens we explored end of last year. I believe the knowledge of when to use which is communicated between dev and design.

> By exporting 4X the number of icons we need to be, ineffective treeshaking results in bombing bundle size.

The library is setup to be tree-shaken and one can confirm this in a fresh install of CRA. Times when this deopts are due to bundler config/settings/versions. 

![image](https://user-images.githubusercontent.com/3901764/73083853-86bdff00-3e91-11ea-9136-4ae9f479e8cf.png)

If you have steps to show how this deopts let me know 👍 

---

> but these are scaleable vector graphics, they shouldn’t be bespoke sprites at individual sizes.

I think one goal of the project would be to only maintain one asset per icon. What was observed was that certain assets rendered poorly on a wide breadth of devices/screens when drawn on 16px artboards. I believe this is what caused the situation that we have today. Could you share how this problem could be addressed?

**Comment by vpicone** [comment]
blocked by https://github.com/carbon-design-system/carbon/issues/5170

**Comment by vpicone** [comment]
@designertyler it looked like you were about to drop a 🔥 spec for this. 

https://github.com/carbon-design-system/carbon-website/issues/766#issuecomment-577749730

@laurenmrice did you have something for it?

**Comment by designertyler** [comment]
@vpicone This is based on specs we had a while ago. I'd wait for the 👍 from @mjabbink and @jeanservaas before you start slinging code.

![Icon page components](https://user-images.githubusercontent.com/19270502/73946824-27fa7b80-48bc-11ea-905c-9ebba7ad565a.png)

**Comment by chrisconnors-ibm** [comment]
What does “copy” do here? Ask was “code snippet to summon from package”, but that’s not clear. And at what size would you get?

**Comment by chrisconnors-ibm** [comment]
Proposal:
1. **Programatic Name** on card (ie `add--filled`) - make it clickable to put the programatic name on the clipboard
2. Menu: “**Copy component code**” - component code to clipboard (NOT THE RAW SVG)
3. Menu: “**Download**” - direct download of 32 asset (from the repo ie https://raw.githubusercontent.com/carbon-design-system/carbon/master/packages/icons/svg/32/badge.svg)
4. Menu: **View in repo**” link to asset repo (if they want the raw SVG they can get it here, the friction to do so should reduce the likelihood **ON PURPOSE**)

The big unanswered question is how to express the size prop for 2 (this is distinct from what the package does with that prop).

The Icon Library page should also have a link to the AI Master; esp if a search result is 0
“No icons found; if you’d like to create (and submit) an icon asset, base the work on the assets in [this master file](https://github.com/carbon-design-system/carbon/blob/master/packages/icons/master/ui-icon-master.ai?raw=true).”

**Comment by jeanservaas** [comment]
Thanks @designertyler the above spec is correct @vpicone with a couple of modifications. No need to look at Chiu's old hex codes from the spec, those are old just pull in the overflow component from Carbon.

In order to make sure that there is no truncation in the overflow menu at 1056 we are going to ditch the icons.

![image](https://user-images.githubusercontent.com/43144260/74055048-daa60900-49a4-11ea-8796-ab7c6185b0bb.png)


There's a battle raging over the content of the dropdown... @vpicone has proposed that the menu say:

- Download SVG
- React

**Anna, Jan and Lauren made several good points against this:**

1. React is our default therefore any code you're copying from Carbon is react — we don't say react before every code snippet on the website

2. We need a action verb, not just a proper noun with no explanation

**Proposed change:**

**- Download SVG**

**- Copy code**

**Comment by mjabbink** [comment]
need to clarify what “Copy code” is doing. I would assume it’s react if only one option presented. Do we need to even do it for developers? Could it only be a download asset action? That seems to be the use case/audience. Have we had devs wanting to Copy React? Let’s only do this for the folks that are clamoring for it. I have only heard this from designers and non designers.

**Comment by chrisconnors-ibm** [comment]
@jnm2377's main asks here were:

1. display the programatic name like  `add-filled`
2. get the react code onto the clipboard

The programatic name is important for designers too since they need that value to specify which icon FED should use in a spec.

She can confirm.

**Comment by chrisconnors-ibm** [comment]
> - **Download SVG**

There's no need to articulate the file type (svg). If you select 'Download', a file appears in your downloads folder.

**Comment by mjabbink** [comment]
Don’t you think people want to know what they’re downloading? I certainly do. I see more reason to keep it than to remove. It’s clear and not harmful or ugly.

**Comment by chrisconnors-ibm** [comment]
I think if it's not obvious that you're going to get an image file for the icon, it will be when it shows up in your Downloads folder 2 seconds later :) Does the he actual filetype (svg vs .png etc.) really matter at that point?

**Comment by vpicone** [comment]
Currently the carbon react OverflowMenu breaks the page if there's more than a few hundred. Implementation of this behavior is going to be slowed by either creating our own or waiting for a fix upstream from Carbon.

Added on to this issue: https://github.com/carbon-design-system/carbon/issues/5079#issuecomment-584469193

**Comment by chrisconnors-ibm** [comment]
we could forgo the menu and have icon buttons on each card. the obvious downside is that at the smallest cardsize, it gets pretty busy. but at least it'd be there until the issue can be unblocked.
![01 Icon Library](https://user-images.githubusercontent.com/35537391/74284600-0d783600-4cf2-11ea-8ebd-ab732a5723bb.png)

when the user hovers over the code icon, display the react name in a tooltip:
![02 Hover - display react name tip](https://user-images.githubusercontent.com/35537391/74284630-1ff26f80-4cf2-11ea-9f69-ee3c98d12fb4.png)

if they click on the code icon, copy the react name to the clipboard and confirm in the tooltip:
![03 Click - confirm copy](https://user-images.githubusercontent.com/35537391/74284669-36003000-4cf2-11ea-943e-c5cad1d0f328.png)

if the hover over the dl icon, tell them what what the button does in a tooltip:
![04 Hover - download SVG tip](https://user-images.githubusercontent.com/35537391/74284699-4b755a00-4cf2-11ea-813a-cd415e78c5a2.png)

if they click the dl icon, confirm their download in the tooltip:
![05 Click - confirm download](https://user-images.githubusercontent.com/35537391/74284725-5f20c080-4cf2-11ea-955f-9f49a87fda4a.png)

**Comment by mjabbink** [comment]
Could work. a bit busy but the most straight forward and immediate

**Comment by mjabbink** [comment]
@vpicone confirming you solved for overflow approach to work. That one seems to be best strategy because it encourages a more considered download behavior or decision.

**Comment by chrisconnors-ibm** [comment]
@mjabbink @jeanservaas even if the two actions go in that menu, the spec doesn't yet cover how/where to display the react name.

**Comment by aagonzales** [comment]
Are you talking about the path name? Doesn't that depend on which size is being called.

**Comment by chrisconnors-ibm** [comment]
no - the asset name itself.
![image](https://user-images.githubusercontent.com/35537391/74373205-01e84600-4daa-11ea-9043-958650881074.png)

**Comment by aagonzales** [comment]
- Why does the download tooltip say the action but the code tooltip say the name and not the action that will be preformed? 
- What is the user getting when the click react code? Just the name, a path?
- Is this supposed to be an MVP? Whatever happened to whole icon drawer that Shixie made?

**Comment by aagonzales** [comment]
For what it's worth I like this solution. Its clean and keeps the users attention on the actual icons. The action icons upfront though nice for quicker access, I think make it harder to visual scan across the page. 

![image](https://user-images.githubusercontent.com/11670886/74375227-94d2b180-4da5-11ea-82ad-c8c910c6ced6.png)

**Comment by mjabbink** [comment]
Me too. Go with this.

Last question is if SVG is cap and does React need initial cap?

**Comment by chrisconnors-ibm** [comment]
Any menu-based proposal depends on this issue being fixed:
https://github.com/carbon-design-system/carbon/issues/5079#issuecomment-584469193

It's been open for 9 days, and currently in the backlog unassigned. If the answer is do nothing until we can deliver a menu based approach, that delivery is an indeterminate date.

**Comment by jeanservaas** [comment]
Hey icon fanatics!!!

and @vpicone 

After a long talk with Vince, Anna, mike and Lauren and I this is the new solution / spec that we landed on

<img width="367" alt="image" src="https://user-images.githubusercontent.com/43144260/74381439-124fef00-4db1-11ea-8709-c0d411e468de.png">


Once react code is copied to clipboard you will get a tooltip verification — for download this is not needed

<img width="325" alt="image" src="https://user-images.githubusercontent.com/43144260/74381319-d3219e00-4db0-11ea-9342-4d79805768b5.png">

**Comment by chrisconnors-ibm** [comment]
How will someone see the react name? (Without pasting it somewhere)

**Comment by aagonzales** [comment]
**Plot twist** here's the alternative design we're going to try since IDL will only have download and it weird to only have on item in and overflow menu. This functionality mimics what we did in [v9](https://v9.carbondesignsystem.com/guidelines/iconography/library). The action will not be active on mobile because that use-case seems invalid if not rare. 

1. No actions icon are shown by default
2. When a user hovers or focuses on an icon tile, download and copy functions will appear.
3. When hovering/focusing on an individual action icon a tooltip will appear.

![image](https://user-images.githubusercontent.com/11670886/74384732-960cda00-4db7-11ea-8e73-fcc3acb749ec.png)

**Comment by chrisconnors-ibm** [comment]
plot twists are the best

**Comment by jeanservaas** [comment]
Plot twist number two... 

- Let's move those guys to the bottom — same dimensions and functionality; even with wrapping names at 1056, the top of the card is going to get napalmed with activity on hover

- also, Anna's spec doesn't show it, but we'd need to move the names up to have only an 8px separation from the top, making the margin uneven

- the original IDL had them at the bottom on hover I believe and CPC is never wrong

- yes, after 48 hours of jibber jabber this is basically just Chris's spec on hover

<img width="706" alt="image" src="https://user-images.githubusercontent.com/43144260/74395328-ab443180-4dd4-11ea-88b2-ef7f8431559b.png">

**Comment by vpicone** [comment]
In the icon meeting I suggested this approach using the react 'atom' logo. We might want to revisit that considering we have 3 other frameworks that use angle brackets. @chrisconnors-ibm can you spin up a react icon for us? 2px stroke width please.

**Comment by mjabbink** [comment]
@conradennis can spin up the react atom icon

**Comment by mjabbink** [comment]
Confirming IDL download only and in pictogram library as well

**Comment by chrisconnors-ibm** [comment]
@conradennis if the react icon meant to live on the bottom right of the card, it needs to scale cleanly down to 16 x 16.

To clarify @mjabbink's comment I think he's saying for the IDL Icon Library and Pictogram library pages (both sights) they should only offer the dl icon.

Finally maybe this effort is a good opportunity to get the IDL icon/pictogram library pages and the carbon equivalents identical (filtering controls differ in style and geometry right now) 🤞

**Comment by mjabbink** [comment]
@chrisconnors-ibm good point. I have an issue to simplify the hierarchy of the icons to be more flat like the Pictograms were done. Too many headers. We can just keep the top level.
https://app.zenhub.com/workspaces/auxiliary-squad-5deea70733c75ceaf1bfacae/issues/carbon-design-system/design-language-website/348

**Comment by mjabbink** [comment]
@conradennis we should se the React icon and download icons side by side in design situation

**Comment by mjabbink** [comment]
This is my hack to see what it would look like

<img width="703" alt="Screen Shot 2020-02-14 at 3 13 19 PM" src="https://user-images.githubusercontent.com/32881239/74568009-af8b5e80-4f3c-11ea-9a6b-6cf5a89bd5bf.png">

**Comment by mjabbink** [comment]
Can’t we assume </> will work in Carbon site?

**Comment by chrisconnors-ibm** [comment]
@mjabbink  - I am referring to the search and filter controls. 

IDL:
![image](https://user-images.githubusercontent.com/35537391/74570962-2deefd00-4f4b-11ea-8ac9-f819e81eaa45.png)

CDS:
![image](https://user-images.githubusercontent.com/35537391/74571003-3ba48280-4f4b-11ea-8e18-95c63d76ec18.png)

**Comment by chrisconnors-ibm** [comment]
The headings help people figure out where they live in the plugin:
![image](https://user-images.githubusercontent.com/35537391/74571196-7dcdc400-4f4b-11ea-87e1-d8a80d4c93db.png)

We could investigate other ways to convey that though.

**Comment by chrisconnors-ibm** [comment]
@mjabbink 
> Can’t we assume </> will work in Carbon site?

I think in a meeting I wasn't part of @vpicone @jeanservaas maybe others thought if we use the react icon people won't be surprised that what they got was react code.

I was ok with `</>`.

**Comment by mjabbink** [comment]
I’m ok with </> and prefer it over react atom unless that is really more clear.

**Comment by mjabbink** [comment]
Also we went UI icons to be clear since we have several types of iconography. UI icons, app icons and Pictograms

**Comment by chrisconnors-ibm** [comment]
this behavior not present @ [preview](https://carbon-website-git-fork-vpicone-icon-lib-action-bar.carbon-design-system.now.sh/guidelines/icons/library)
https://github.com/carbon-design-system/carbon-website/issues/766#issuecomment-585460128
Specifically

>When hovering/focusing on an individual action icon a tooltip will appear.
>"Download SVG"
>"Copy <icon-name />"

cc @aagonzales @jeanservaas

**Comment by aagonzales** [comment]
I think now that it has been deployed any outstanding issues or bugs need to be opened in a new issue.

**Comment by mjabbink** [comment]
This is meant for IDL as well and Pictograms

**Comment by aagonzales** [comment]
[this issue that will never be closed]

FYI this is the carbon-website repo, it may need to be moved or a new issue opened for IDL.

**Comment by vpicone** [comment]
@mjabbink the PR you just closed in IDL is adding them there (https://github.com/carbon-design-system/design-language-website/pull/439)

Pictograms _were_ included in the initial PR and are live.

**Comment by chrisconnors-ibm** [comment]
I thought Pictograms (IDL and CDS) and Icon cards (IDL) were only supposed to get the DL button:
https://github.com/carbon-design-system/carbon-website/issues/766#issuecomment-585543843

**Comment by mjabbink** [comment]
Yes, download only

**Comment by vpicone** [comment]
@mjabbink @chrisconnors-ibm It seems like pictograms would be _more_ requested for use in keynote/outside of our libraries. Is there a reason we wouldn't want folks to download it?

**Comment by mjabbink** [comment]
@vpicone yes, pictograms and icons are to have download functionality

**Comment by chrisconnors-ibm** [comment]
>I thought Pictograms (IDL and CDS) and Icon cards (IDL) were **only** supposed to get the DL button:
https://github.com/carbon-design-system/carbon-website/issues/766#issuecomment-585543843

At issue is whether either IDL Pictogram cards or IDL Icon Cards get the code `</>` action

**Comment by mjabbink** [comment]
go for it.

**Comment by mjabbink** [comment]
I thought that saying DL only for icons and pictograms was clear. But if you all think </> is useful in Idl for pictograms then do it. I was thinking it was not but I based that on my own opinion only.

**Comment by mjabbink** [comment]
Could have both really even if the </> is hardly ever used.

**Comment by vpicone** [comment]
@chrisconnors-ibm we have the `pictograms-react` library, not sure why we wouldn't have the copy code button for them on CDS.

Right now CDS has both buttons for both libraries, IDL has just download for both libraries. Not sure what the issue is?

**Comment by mjabbink** [comment]
We can have the libraries be identical I guess. Not harm to have copy code in IDL and I’m all for making them be exactly the same.
